### ISSUE TITLE
[Question] Using a trusted certificate 

### ISSUE THREAD

**Issue by ClearGateTeam** [n1nj4sec/pupy#460]
I'm trying to use a purchased certificate instead of the self signed certificate being generated by Pupy.

What changes should be taken in order for that to work?

**Comment by alxchk** [comment]
Not an easy question. The logic in ssl server should be changed to not to check the role of the neighbor. Or better to add some kind of SNI checking to use different certs for https and control connection. Originally there is a check that ensures that neighbor role is not the same as yours. Cause in case of one client compromise youll be able to perform mitm attack to the rest clients which is not good.
Are you really need this during pentest?

**Comment by ClearGateTeam** [comment]
yes. because some customers are using firewall that checks for known CA for HTTPS sessions.

Is there a way to disable the default behavior?

**Comment by alxchk** [comment]
The logic is in pupy/network/lib/clients.py - PupySSLClient and
pupy/network/transports/ssl/conf.py - PupySSLAuthenticator (server side).
You can try to tune it according to your needs. Just remember to rebuild templates.

**Comment by ClearGateTeam** [comment]
What should be modified in order to completely disable the check?

**Comment by alxchk** [comment]
You can try to start with disabling client_cert_required:
`config -w set ssl client_cert_required false`

Then, if it's not enough you can try to remove all checks completely with replacing
`cert_reqs=self.cert_reqs` with `cert_reqs=ssl.CERT_NONE`

You can check python ssl module documentation in case something will not work (https://docs.python.org/2/library/ssl.html)

**Comment by ClearGateTeam** [comment]
thanks @alxchk .

I'm now trying to rebuild the template but got some errors like:

`/root/pupy-new/pupy/client/sources/buildenv/win64/python.sh ../build_library_zip.py -windows
Traceback (most recent call last):
  File "../build_library_zip.py", line 11, in <module>
    import pp
  File "..\..\pupy\pp.py", line 59, in <module>
    from network import conf
  File "..\..\pupy\network\conf.py", line 6, in <module>
    from .lib.launchers.connect import ConnectLauncher
  File "..\..\pupy\network\lib\__init__.py", line 11, in <module>
    from .transports.aes import AES256, AES128
  File "..\..\pupy\network\lib\transports\aes.py", line 15, in <module>
    from Crypto.Protocol.KDF import PBKDF2
  File "C:\Python27\lib\site-packages\Crypto\Protocol\KDF.py", line 39, in <module>
    from Crypto.Hash import SHA1, SHA256, HMAC, CMAC
  File "C:\Python27\lib\site-packages\Crypto\Hash\SHA256.py", line 39, in <module>
    from Crypto.Util._raw_api import (load_pycryptodome_raw_lib,
  File "C:\Python27\lib\site-packages\Crypto\Util\_raw_api.py", line 39, in <module>
    ffi = FFI()
  File "C:\Python27\lib\site-packages\cffi\api.py", line 63, in __init__
    from . import cparser
  File "C:\Python27\lib\site-packages\cffi\cparser.py", line 7, in <module>
    import pycparser
  File "C:\Python27\lib\site-packages\pycparser\__init__.py", line 14, in <module>
    from .c_parser import CParser
  File "C:\Python27\lib\site-packages\pycparser\c_parser.py", line 20, in <module>
    class CParser(PLYParser):
  File "C:\Python27\lib\site-packages\pycparser\plyparser.py", line 95, in template
    _create_param_rules(cls, method)
  File "C:\Python27\lib\site-packages\pycparser\plyparser.py", line 112, in _create_param_rules
    param_rule.__doc__ = func.__doc__.replace('xxx', xxx).replace('yyy', yyy)
AttributeError: 'NoneType' object has no attribute 'replace'
Makefile:71: recipe for target 'resources/library.zip' failed
make: *** [resources/library.zip] Error 1
`

Is there any recommended distribution to perform this action ?

**Comment by alxchk** [comment]
It's because somehow you have PyCParser version which doesn't support -OO. You need version 2.17. Check requirements.txt. Or better pull from unstable branch, remove environments and build windows one again.

**Comment by ClearGateTeam** [comment]
I did so but still not working. when setting --lvl DEBUG, I see:

> failed to authenticate, rejecting connection

**Comment by alxchk** [comment]
This means ssl_authenticator (pupy/network/transports/ssl/conf.py) failed. Ensure client_cert_required is set to false, cert_reqs set to ssl.CERT_NONE. Because this works on pupysh side you don't need to rebuild anything, just restart pupysh. Feel free to add prints to check that these variables set to expected values.

**Comment by ClearGateTeam** [comment]
@alxchk big thanks.

It finally worked. I had to comment out below lines in clients.py:

` for (item) in peer['subject']:
            if item[0][0] == 'organizationalUnitName':
                peer_role = item[0][1]

        if not ( self.ROLE == 'CLIENT' and peer_role == 'CONTROL' or \
          self.ROLE == 'CONTROL' and peer_role == 'CLIENT' ):
          raise ValueError('Invalid peer role: {}'.format(peer_role))`

and changed the value found above the code from `peer_role = ''` to `peer_role = ''CLIENT`
I did it because in my certificate the subject field does not contain the strings CLIENT or CONTROL.

From your experience, is this modification can cause any issues?

In addition, should I also comment out the same code found in conf.py?
